Unrecognized associated type bound on another associated type

``` rust
trait Bar<T> { fn dummy(&self); }
trait Car<T> { fn dummy(&self); }

trait Foo {
    type A;
    type B: Bar<Self::A>;
    type C: Car<Self::A>;

    fn get_b(&self) -> &Self::B;
}

fn test_bar<A, B: Bar<A>>(_: &B) {}

fn test<A, F: Foo<A=A>>(f: &F) {
    test_bar(f.get_b());
}
```

Gives me:

```
<anon>:15:16: 15:23 error: the trait `Bar<A>` is not implemented for the type `<F as Foo>::B` [E0277]
<anon>:15     test_bar(f.get_b());
                         ^~~~~~~
<anon>:15:16: 15:23 error: the trait `Car<A>` is not implemented for the type `<F as Foo>::C` [E0277]
<anon>:15     test_bar(f.get_b());
                         ^~~~~~~
error: aborting due to 2 previous errors
playpen: application terminated with error code 101
```

There is a workaround:

``` rust
fn test<A, B: Bar<A>, C: Car<A>, F: Foo<A=A, B=B, C=C>>(f: &F) {
    test_bar(f.get_b());
}
```

But it's ugly and should not be necessary.

We hit this problem a lot with [gfx::Device](https://github.com/gfx-rs/gfx-rs/blob/68298c260b8d420c9c1cfc8657c87462a84476a6/src/device/mod.rs#L233), and I'd like to see cleaner use of it without explicit `CommandBuffer` bounds that we use as a [workaround](https://github.com/PistonDevelopers/gfx-debug-draw/blob/58e6d6206a0632492723566ce018ce5a85536c35/src/text_renderer.rs#L246).

This is the error now:

```


rustc 1.13.0 (2c6933acc 2016-11-07)
error[E0405]: trait `std::marker::PhantomFn` is not in scope
 --> <anon>:1:15
  |
1 | trait Bar<T>: std::marker::PhantomFn<T> { fn dummy(&self); }
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^ `std::marker::PhantomFn` is not in scope
  |
  = help: no candidates by the name of `PhantomFn` found in your project; maybe you misspelled the name or forgot to import an external crate?

error[E0405]: trait `std::marker::PhantomFn` is not in scope
 --> <anon>:2:15
  |
2 | trait Car<T>: std::marker::PhantomFn<T> { fn dummy(&self); }
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^ `std::marker::PhantomFn` is not in scope
  |
  = help: no candidates by the name of `PhantomFn` found in your project; maybe you misspelled the name or forgot to import an external crate?

error: main function not found

error: cannot continue compilation due to previous error

```

Edited original post to remove `PhantomFn` bounds, they're no longer needed. Issue still reproduces:

```
error[E0277]: the trait bound `<F as Foo>::B: Bar<A>` is not satisfied
  --> test.rs:14:1
   |
14 | / fn test<A, F: Foo<A=A>>(f: &F) {
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Bar<A>` is not implemented for `<F as Foo>::B`
   |
   = help: consider adding a `where <F as Foo>::B: Bar<A>` bound
   = note: required by `Foo`

error[E0277]: the trait bound `<F as Foo>::C: Car<A>` is not satisfied
  --> test.rs:14:1
   |
14 | / fn test<A, F: Foo<A=A>>(f: &F) {
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Car<A>` is not implemented for `<F as Foo>::C`
   |
   = help: consider adding a `where <F as Foo>::C: Car<A>` bound
   = note: required by `Foo`

error: aborting due to 2 previous errors
```

Ran into this as well. Here's my minimal test case:

```
trait Bar<T> {}
trait Caz {
    type A;
    type B: Bar<Self::A>;
}

fn test<T, U>(_: T) where T: Caz, U: Caz<A = T::A> {}
```

```text
error[E0277]: the trait bound `<U as Caz>::B: Bar<<T as Caz>::A>` is not satisfied
 --> src/main.rs:7:1
  |
7 | fn test<T, U>(_: T) where T: Caz, U: Caz<A = T::A> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Bar<<T as Caz>::A>` is not implemented for `<U as Caz>::B`
  |
  = help: consider adding a `where <U as Caz>::B: Bar<<T as Caz>::A>` bound
note: required by `Caz`
 --> src/main.rs:2:1
  |
2 | trait Caz {
  | ^^^^^^^^^
```

[Current output](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=90590bd4d71c4c30ca1cd791a1d45870):

```
error[E0277]: the trait bound `<F as Foo>::B: Bar<A>` is not satisfied
  --> src/lib.rs:14:1
   |
4  |   trait Foo {
   |   --------- required by `Foo`
...
14 |   fn test<A, F: Foo<A=A>>(f: &F) {
   |   ^                             - help: consider further restricting the associated type: `where <F as Foo>::B: Bar<A>`
   |  _|
   | |
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Bar<A>` is not implemented for `<F as Foo>::B`

error[E0277]: the trait bound `<F as Foo>::C: Car<A>` is not satisfied
  --> src/lib.rs:14:1
   |
4  |   trait Foo {
   |   --------- required by `Foo`
...
14 |   fn test<A, F: Foo<A=A>>(f: &F) {
   |   ^                             - help: consider further restricting the associated type: `where <F as Foo>::C: Car<A>`
   |  _|
   | |
15 | |     test_bar(f.get_b());
16 | | }
   | |_^ the trait `Car<A>` is not implemented for `<F as Foo>::C`
```

[Applying the suggestions](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e43b5d6a74a5815b625691808420f3a0) makes the code compile.

---

[Niko's minimized version](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=056093e1739899755b3f962da894dc35):

```
error[E0277]: the trait bound `<F as Foo>::B: Bar<A>` is not satisfied
  --> src/main.rs:12:1
   |
3  |   trait Foo {
   |   --------- required by `Foo`
...
12 |   fn test<A, F: Foo<A=A>>(f: &F) {
   |   ^                             - help: consider further restricting the associated type: `where <F as Foo>::B: Bar<A>`
   |  _|
   | |
13 | |     test_bar(f.get_b());
14 | | }
   | |_^ the trait `Bar<A>` is not implemented for `<F as Foo>::B`
```

---

@daboross' [case](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7afcdab09435cf3010f0c985a5aaa0b6):

```
error[E0277]: the trait bound `<U as Caz>::B: Bar<<T as Caz>::A>` is not satisfied
 --> src/lib.rs:7:1
  |
2 | trait Caz {
  | --------- required by `Caz`
...
7 | fn test<T, U>() where T: Caz, U: Caz<A = T::A> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-^^
  | |                                             |
  | |                                             help: consider further restricting the associated type: `, <U as Caz>::B: Bar<<T as Caz>::A>`
  | the trait `Bar<<T as Caz>::A>` is not implemented for `<U as Caz>::B`
```

---

In call cases applying the suggestions fulfill the requirements.

--- a/src/test/ui/associated-types/issue-24159.rs
+++ b/src/test/ui/associated-types/issue-24159.rs
@@ -1,0 +1,37 @@
+// check-pass
+
+#![allow(unused)]
+
+trait Bar<T> {
+    fn dummy(&self);
+}
+
+trait Foo {
+    type A;
+    type B: Bar<Self::A>;
+
+    fn get_b(&self) -> &Self::B;
+}
+
+fn test_bar<A, B: Bar<A>>(_: &B) {}
+
+fn test<A, F: Foo<A = A>>(f: &F) {
+    test_bar(f.get_b());
+}
+
+trait Bar1<T> {}
+trait Caz1 {
+    type A;
+    type B: Bar1<Self::A>;
+}
+
+fn test1<T, U>() where T: Caz1, U: Caz1<A = T::A> {}
+
+trait Bar2<T> {}
+trait Caz2 {
+    type A;
+    type B: Bar2<Self::A>;
+}
+fn test2<T: Caz2<A = ()>>() {}
+
+fn main() {}

--- a/src/test/ui/associated-types/issue-37808.rs
+++ b/src/test/ui/associated-types/issue-37808.rs
@@ -1,0 +1,19 @@
+// check-pass
+
+trait Parent {
+    type Ty;
+    type Assoc: Child<Self::Ty>;
+}
+
+trait Child<T> {}
+
+struct ChildWrapper<T>(T);
+impl<A, T> Child<A> for ChildWrapper<T> where T: Child<A> {}
+
+struct ParentWrapper<T>(T);
+impl<A, T: Parent<Ty = A>> Parent for ParentWrapper<T> {
+    type Ty = A;
+    type Assoc = ChildWrapper<T::Assoc>;
+}
+
+fn main() {}

--- a/src/test/ui/associated-types/issue-37883.rs
+++ b/src/test/ui/associated-types/issue-37883.rs
@@ -1,0 +1,25 @@
+// check-pass
+
+use std::ops::Mul;
+
+fn main() {}
+
+trait Ring {}
+trait Real: Ring {}
+
+trait Module: Sized + Mul<<Self as Module>::Ring, Output = Self> {
+    type Ring: Ring;
+}
+
+trait EuclideanSpace {
+    type Coordinates: Module<Ring = Self::Real>;
+    type Real: Real;
+}
+
+trait Translation<E: EuclideanSpace> {
+    fn to_vector(&self) -> E::Coordinates;
+
+    fn powf(&self, n: <E::Coordinates as Module>::Ring) -> E::Coordinates {
+        self.to_vector() * n
+    }
+}

--- a/src/test/ui/associated-types/issue-39532.rs
+++ b/src/test/ui/associated-types/issue-39532.rs
@@ -1,0 +1,14 @@
+// check-pass
+
+#![allow(unused)]
+
+trait Foo {
+    type Bar;
+    type Baz: Bar<Self::Bar>;
+}
+
+trait Bar<T> {}
+
+fn x<T: Foo<Bar = U>, U>(t: &T) {}
+
+fn main() {}

